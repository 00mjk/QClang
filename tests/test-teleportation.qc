tup(qubit, qubit) epr_pair()
{
	return CX(hadamard(false), false);
}

tup(bit, bit) measure_bell(qubit a, qubit b)
{
	(a, b) = CX(a, b);
	return (measure(hadamard(a)), measure(b));
}

qubit teleport(qubit alice)
{
	qubit shared;
	qubit bob;
	bit meas1;
	bit meas2;

	(shared, bob) = epr_pair();
	(shared, bob) = barrier(shared, bob);
	/* enabling this causes the output to be random, which our test script can't handle atm */
	/* (meas1, meas2) = measure_bell(alice, shared); */

	/* TODO invert based on measurement */
	return bob;
}

int main()
{
	qubit alice;
	qubit bob;

	alice = true;
	bob = teleport(alice);

	return 0;
}
