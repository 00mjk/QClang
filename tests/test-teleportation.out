OPENQASM 2.0;
include "qelib1.inc";
qreg q[1];
creg c[1];
h q;
qreg temp_0[1];
x temp_0;
qreg temp_3[1];
h temp_3;
qreg temp_4[1];
cx temp_3, temp_4;
barrier temp_3, temp_4;
